### ISSUE TITLE
kube-apiserver: very high memory load in Raspberry PI

### ISSUE THREAD

**Issue by Bruce17** [ubuntu/microk8s#1900]
Hi,

I'm running a freshly installed MicroK8s cluster v1.20 on top of Ubuntu Server 20.04 on 3 Raspberry PI v4. After a couple of hours I noticed that some of my pods are not responing anymore and kubectl doesn't return anything either. A look onto the system reveled the issue: full memory and the process kube-apiserver took around 70-80 % of it.

![image](https://user-images.githubusercontent.com/5036995/104697999-ed76a100-5710-11eb-9a39-6f6e69d8d94f.png)

At the moment I execute `sudo snap restart microk8s.daemon-apiserver` every hour (cronjob) to keep the cluster somehow alive.

Recently I switched from K3s to MicroK8s due to high cpu and memory load in K3s. Maybe there is an overall issue with K8s on top of Raspberry PIs?

Targerball of `microk8s inspect`:
[inspection-report-20210115_091600.tar.gz](https://github.com/ubuntu/microk8s/files/5819221/inspection-report-20210115_091600.tar.gz)

**Comment by Ramblurr** [comment]
I am experiencing the same problem. Every morning I wake up and my microk8s cluster is hung, with apiserver pegging the cpu and all ram consumed and logs filled with messages from linux oom killer.

**Comment by balchua** [comment]
Hi @Bruce17 i believe what you are experiencing is similar to https://github.com/ubuntu/microk8s/issues/1598
For now some users switch to non HA.
`microk8s disable ha-cluster`.
You will have to do this to all nodes and then join them back together.
Remember switching from HA to non HA might lose your data.

**Comment by Bruce17** [comment]
Hi @balchua,
thx for the hint. The default memory load went down from ~1.4 GB to around 800 MB on the master node and ~430 MB on the nodes. I will keep an eye on that the next days.

Also I disabled the crontab entry to automatically restart `microk8s.daemon-apiserver` for the moment.

A question to get it clear: disabling HA does only disable the HA feature of the master nodes, right? Will load balancing inside of the Kubernetes cluster e.g. if node A goes down move pods to node B does still work?

**Comment by balchua** [comment]
Yes pods which were previously on nodeA will be moved to nodeB when nodeB goes down.  I think its like 5 minutes (configured through controller manager).
The only thing is if the controlplane goes down, there's no way to access the cluster.  Not the best way to run a prod cluster though.

**Comment by Bruce17** [comment]
A noticed a couple of things. Here are my findings after disabling HA:

1. Removing nodes from the cluster and disabling HA did not solve the problem. I was not able to properly get the cluster up and running again though a `microk8s reset --destroy-storage` was required.
2. After that there were still issues with flanneld. My nfs-client was not able to connect to the master node to gather a nfs pvc.
3. Complete fresh setup of microk8s using v1.19 this time with disabled HA.
Nfs was working this time, but I was not able to deploy a pod (except k8s dashboard - weird). It was failing with:
`Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "869f6cbee311e603bc959ea086560386899d38545f435a039cd4ff86b0cc9387": error getting ClusterInformation: Get https://[10.152.183.1]:443/apis/crd.projectcalico.org/v1/clusterinformations/default: x509: certificate signed by unknown authority (possibly because of "crypto/rsa: verification error" while trying to verify candidate authority certificate "10.152.183.1")`
Seems a node was not able to connect to the master. Found another one having a same [issue](https://discuss.kubernetes.io/t/solved-x509-certificate-error/14151/3), but no proper solution either.
4. Uninstalling microk8s again using: `snap remove microk8s --purge`
5. Installing microk8s v1.20.1 with disabled HA. Seems now it is working, but I will monitor it again over the next days,

**Comment by rockkoca** [comment]
This is a memory leak bug. Try to monitor the memory usage and restart the API server.

```
sudo systemctl status snap.microk8s.daemon-apiserver
sudo systemctl restart snap.microk8s.daemon-apiserver
```

**Comment by Ramblurr** [comment]
For folks landing here from Google, this seems to be the current summary:

* There's a memory leak when HA mode is enabled
* There is no solution or workaround that lets you keep using HA mode
* Disable HA mode might fix it, but you might have to purge and reinstall

**Comment by balchua** [comment]
To those who are interested, tthere are few fixes on dqlite.  The fixes will be backported to 1.19. and 1.20.

https://github.com/ubuntu/microk8s/issues/1598#issuecomment-779044844
